File: swarm-manage/internal/core/domain/swarm.go
package domain

import "time"

// ─── Stacks ──────────────────────────────────────────────────────────────────

type Stack struct {
	Name      string    `json:"name"`
	Services  int       `json:"services"`
	CreatedAt time.Time `json:"createdAt"`
}

// ─── Services ─────────────────────────────────────────────────────────────────

type ServiceReplicas struct {
	Running uint64 `json:"running"`
	Desired uint64 `json:"desired"`
}

type Service struct {
	ID       string          `json:"id"`
	Name     string          `json:"name"`
	Image    string          `json:"image"`
	Stack    string          `json:"stack"`
	Replicas ServiceReplicas `json:"replicas"`
	UpdatedAt time.Time      `json:"updatedAt"`
}

// ─── Nodes ────────────────────────────────────────────────────────────────────

type Node struct {
	ID           string `json:"id"`
	Hostname     string `json:"hostname"`
	Role         string `json:"role"`
	Status       string `json:"status"`
	Availability string `json:"availability"`
	EngineVersion string `json:"engineVersion"`
}

// ─── Stats ────────────────────────────────────────────────────────────────────

// ContainerStats is a one-shot resource snapshot for a single swarm-managed
// container running on this node. cpuPercent comes from the one-shot precpu
// sample; net/blk values are cumulative byte counters since container start.
type ContainerStats struct {
	ContainerID string  `json:"containerId"`
	TaskID      string  `json:"taskId"`
	NodeID      string  `json:"nodeId"`
	ServiceID   string  `json:"serviceId"`
	ServiceName string  `json:"serviceName"`
	Stack       string  `json:"stack"`
	State       string  `json:"state"`
	CPUPercent  float64 `json:"cpuPercent"`
	MemUsage    uint64  `json:"memUsage"`
	MemLimit    uint64  `json:"memLimit"`
	NetRx       uint64  `json:"netRx"`
	NetTx       uint64  `json:"netTx"`
	BlockRead   uint64  `json:"blockRead"`
	BlockWrite  uint64  `json:"blockWrite"`
	Error       string  `json:"error,omitempty"`
}

File: app/src/types/swarm.ts
export interface Stack {
  name: string;
  services: number;
  createdAt: string;
}

export interface ServiceReplicas {
  running: number;
  desired: number;
}

export interface SwarmService {
  id: string;
  name: string;
  image: string;
  stack: string;
  replicas: ServiceReplicas;
  updatedAt: string;
}

export interface SwarmNode {
  id: string;
  hostname: string;
  role: string;
  status: string;
  availability: string;
  engineVersion: string;
}

// Matches swarm-manage's domain.ContainerStats — cumulative counters since
// container start. One row per swarm-managed container running on the server.
export interface ContainerStats {
  containerId: string;
  taskId: string;
  nodeId: string;
  serviceId: string;
  serviceName: string;
  stack: string;
  state: string;
  cpuPercent: number;
  onlineCpus: number;
  memUsage: number;
  memLimit: number;
  netRx: number;
  netTx: number;
  blockRead: number;
  blockWrite: number;
  error?: string;
}
